File: src/main/java/work/xeltica/craft/core/models/EbiPowerItem.java
package work.xeltica.craft.core.models;

import java.util.LinkedHashMap;
import java.util.Map;

import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

/**
 * エビパワーストアで取り扱う商品を表すモデルです。
 * @author Xeltica
 */
public record EbiPowerItem(ItemStack item, int cost) implements Cloneable, ConfigurationSerializable {

    @Override
    public @NotNull Map<String, Object> serialize() {
        final var result = new LinkedHashMap<String, Object>();
        result.put("item", item.serialize());
        result.put("cost", cost);

        return result;
    }

    public static EbiPowerItem deserialize(Map<String, Object> args) {
        final var item = ItemStack.deserialize((Map<String, Object>) args.get("item"));
        final var cost = (int) args.get("cost");

        return new EbiPowerItem(item, cost);
    }
}

File: src/main/java/work/xeltica/craft/core/modules/ebipowerShop/EbiPowerItem.kt
package work.xeltica.craft.core.modules.ebipowerShop

import org.bukkit.configuration.serialization.ConfigurationSerializable
import org.bukkit.inventory.ItemStack
import work.xeltica.craft.core.utils.CastHelper

/**
 * エビパワーストアで購入可能なアイテムを表す、シリアライズ可能なデータクラス。
 */
data class EbiPowerItem(val item: ItemStack, val cost: Int) : Cloneable, ConfigurationSerializable {
    override fun serialize(): MutableMap<String, Any> {
        val result = LinkedHashMap<String, Any>()
        result["item"] = item.serialize()
        result["cost"] = cost

        return result
    }

    companion object {
        // NOTE: Spigotは動的にこの関数を呼び出すため、JvmStaticでなければならない
        @JvmStatic
        fun deserialize(args: Map<String, Any>): EbiPowerItem {
            val item = ItemStack.deserialize(CastHelper.checkMap<String, Any>(args["item"] as Map<*, *>))
            val cost = args["cost"] as Int

            return EbiPowerItem(item, cost)
        }
    }
}
